src/com/ms/silverking/cloud/zookeeper/ZooKeeperExtended.java

```diff
@@ -42,1024 +43,1081 @@
+    } else {
+      currentVersions = new ArrayList<>(children.size());
+      for (String child : children) {
+        currentVersions.add(Long.parseLong(child));
+      }
+      Collections.sort(currentVersions);
+      return currentVersions;
+    }
+  }
+
+  public long getLatestVersion(String path) throws KeeperException {
+    List<Long> sorted;
+
+    sorted = getSortedVersions(path);
+    return sorted.isEmpty() ? -1 : sorted.get(sorted.size() - 1);
+  }
+
+  public long getLeastVersion(String path) throws KeeperException {
+    List<Long> sorted;
+
+    sorted = getSortedVersions(path);
+    return sorted.isEmpty() ? -1 : sorted.get(0);
+  }
+
+  public long getLatestVersionFromPath(String path) {
+    int index;
+
+    index = path.lastIndexOf('/');
+    return Long.parseLong(path.substring(index + 1));
+  }
+
+  public String getLatestVersionPath(String path) throws KeeperException {
+    return path + "/" + Strings.padStart(Long.toString(getLatestVersion(path)), 10, '0');
+  }
+
+  public String getLeastVersionPath(String path) throws KeeperException {
+    return path + "/" + Strings.padStart(Long.toString(getLeastVersion(path)), 10, '0');
+  }
+
+  public long getVersionPriorTo(String path, long zxid) throws KeeperException {
+    List<String> children;
+    List<Long> currentVersions;
+
+    //System.out.println("getVersionPriorTo "+ path +" "+ zxid);
+    children = getChildrenPriorTo(path, zxid);
+    if (children.size() == 0) {
+      //System.out.println("no version prior to "+ path +" "+ zxid);
+      return -1;
+    } else {
+      currentVersions = new ArrayList<>(children.size());
+      for (String child : children) {
+        currentVersions.add(Long.parseLong(child));
+      }
+      Collections.sort(currentVersions);
+      return currentVersions.get(currentVersions.size() - 1);
+    }
+  }
+
+  public static String parentPath(String childPath) {
+    int lastSlash;
+
+    lastSlash = childPath.lastIndexOf('/');
+    if (lastSlash < 0) {
+      return null;
+    } else {
+      return childPath.substring(0, lastSlash);
     }
-    
-    public static long getTrailingVersion(String path) {
-        int     i;
-        
-        i = path.lastIndexOf('/');
-        if (i < 0) {
-            throw new RuntimeException("Invalid path: "+ path);
-        } else {
-            String  def;
-            
-            def = path.substring(i + 1);
-            try {
-                return Long.parseLong(def);
-            } catch (NumberFormatException nfe) {
-                throw new RuntimeException("Bad version: "+ def);
-            }
-        }
-    }    
+  }
+
+  public static long getTrailingVersion(String path) {
+    int i;
+
+    i = path.lastIndexOf('/');
+    if (i < 0) {
+      throw new RuntimeException("Invalid path: " + path);
+    } else {
+      String def;
+
+      def = path.substring(i + 1);
+      try {
+        return Long.parseLong(def);
+      } catch (NumberFormatException nfe) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)